clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java

```diff
@@ -1263,6 +1265,41 @@ public void testListConsumerGroupsMetadataFailure() throws Exception {
         }
     }
 
+    @Test
+    public void testListConsumerGroupsWithStates() throws Exception {
```

Comment:
Could we add a test case which verifies that `UnsupportedVersionException` is thrown when states are provided but an older API is used? Side question, I haven't seen logic to raise it anywhere. Is it handled implicitly by the serialization layer?

Reply:
Yes, the exception is raised in ListGroupsRequestData. We can't test it here as `MockClient` does not serialize requests. That said, it's a good point, I'll see where we can test this

Reply:
So, it will throw the following exception: `UnsupportedVersionException("Attempted to write a non-default states at version " + _version)`.  The message is a bit weird for an end user. Would it make sense to catch it in `handleFailure` and throw an exception with a better message?

I wonder how such cases have been handled so far. I don't see any particular handling in the admin client at the moment.

Reply:
I just stumbled upon this: https://github.com/apache/kafka/blob/169fa0efcc9089bdcf37e6f8b1b9a6f3377f38f8/clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java#L84

We could perhaps do something similar in our case.

Reply:
I was sort of avoiding doing this, as I believe we should rely on the underlying Data class instead of having such logic in the Request/Response classes. However, For consistency, let's do it and it will also simplify testing for `UnsupportedVersionException`.

Reply:
Thanks. I agree that it is not perfect but it will provide a better error to the end user by explicitly handling it.